### DISCUSSION TITLE
[v3] Re-rendering after wire:click destroys Alpine objects in nested x-for loop

### DISCUSSION THREAD

**Discussion by Enrica-r** [livewire/livewire#6537] [7. Bugs (archived)]
### Livewire version

3.0.1

### Which PHP version are you using?

PHP 8.2

### Repository URL

--

### new knowledge about this issue
First I thought that Livewire 3 destroys properties in Alpine objects. So the last two week I analyzed more and I found out that Livewire 3 does not rerender the whole Alpine objects with nested loops after calling a Livewire function. See my own anwser and a simple code example in next section.

### Steps To Reproduce

1. Clone the above Repo (fresh Laravel installation + Livewire 3.0.1) + install 
2. Open "/" site --> all is OK. Open console --> you will see a date range picker and a button
3. Choose a date range. If you click the second date an event is dispatched to a livewire component. In this moment you will see a bunch of warnings and errors.
4. Disable code line 610 and 611 in file `components/date-range-picker.blade.php` (= $wire.dispatch...)
5. Repeat step 3 and 4 --> it seems OK (but no data is sent to Livewire component)
6. Click button `Dispatch result` which is nothing else than a `wire:click="function"`. --> same bunch of warning and errors of AlpineJS comes up in console.
see example warning: ![console warnings](https://github.com/livewire/livewire/assets/2463365/b0eda0fa-36fc-4951-8609-a8f718ffa578)

### Additional information

The included date range picker component uses AlpineJS and it is based on a Blade component.

The Blade component `components/date-range-picker.blade.php` itself is used within a Livewire component named `search-bar`. After selecting a date range the Blade component dispatches the result to a given name in a property. The listener in this example is in the Livewire `search-bar` component.

After Livewire update request the AlpineJS Object within the Blade component looses a part of his properties. The console log is full of errors.

Even if the code line (610 and 611 in date-range-picker) with `$wire.dispatch` is commented out and you press the button "Dispatch result" which fires a simple click event to a function in `searchbar`, the same errors are fired in console. This button uses a normal `wire:click="functionName"` action.
This means a request from AlpineJS to the Backend destroys my AlpineJS Object in `date-range-picker`.

Same functionallity worked in Livewire 2 with `$wire.emit`

I tried multiple measurements. Eg. replace a computet `get` function by a normale function. Replace the property by a hard coded value. No chance. All properties after `weekdays` are missing.

I have placed this livewire script in head section to avoid dependencies. The Alpine script of component is during development in the blade file but it is pushed once to the bottom of body. 

I can't find a bug on my AlpineJS component by myself. Either I did a non common error or AlpineJS/Livewire 3 has a bug. @calebporzio Please help, I give up!

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by Enrica-r** [comment]
My previous decribed Blade component uses an Alpine object. The rendered HTML code has nested x-for loops. In Livewire 2 the component was rerendered after calling an emit or a wire:click.

With Livewire 3 the rerendering was optimized. If the inner loop is rendered again after calling the backend function, the object is not in memory anymore and Javascript fires a bunch of errors.

I have built a simple stupid example which is perfectly running under Livewire 2. It is a simple list of two weeks with sub-list of days (Monday to Friday). If you click on a day button, the week number plus the day name is showed on the right side and an event is dispatched to Livewire component. Same if you click the button "Toggle shwo Livewire result" which calls a component function with wire:click.

Try with following code:

Component `App\Livewire\ExampleWeekdays`:

``` php
<?php

namespace App\Livewire;

use Livewire\Component;
use Livewire\Attributes\On;
use Livewire\Attributes\Layout;

class ExampleWeekdays extends Component
{


    public string $selectedDay = '';
    public bool $open = false;

    #[On('selectDay')]
    public function selectDay(string $day)
    {
        $this->selectedDay = $day;
    }

    public function toggleShowDay(): void
    {
        $this->open = !$this->open;
    }

    #[Layout('layouts.guest-livewire')]
    public function render()
    {
        return view('livewire.example-weekdays');
    }
}
```

Blade view with Alpine code `views/livewire/example-weekdays.blade.php`:
``` HTML
<div class="max-w-4xl bg-white p-6">

    <h1 class="text-lg mb-6">Example Inline Alpine Component with nested x-for loop and Livewire action outside Alpine div and dispatch inside:
    </h1>
    <div class="mb-2">Click on day name triggers a $dispatch from Alpine to Livewire Component. The result on the right
        side is show in Apline directly.</div>
    <div class="mb-4">Click on Button "Show Result" will fire a wire:click event to Livewire Component</div>

    <div class="mb-2">Result: The inner loop with "weekDays" property fires an Alpine Expression error. The inner loop is not replaced by rerendering. Empty entries are added.</div>
    <div class="mb-4">The error is fired either when Alpine $dispatches or when the wire:click is fired.</div>

    <div x-data="{
        weekDays: [
            'Monday',
            'Tuesday',
            'Wednesday',
            'Thursday',
            'Friday'
        ],
        weeks: [
            {
                weekNo: 1
            },
            {
                weekNo: 2
            }
        ],
        selectedWeekDay: '',
        dayClick(week, dayName) {
            console.log(week, dayName)
            this.selectedWeekDay = week + ' ' + dayName;
            this.$dispatch('selectDay', { day: dayName });
        }
    }">
        <div class="grid grid-cols-[1fr_2fr]">
            <div class="">
                <template x-for="week in weeks" :key="week.weekNo">
                    <div>
                        <div class="flex">
                            <div>Week no.: </div>
                            <div x-text=week.weekNo></div>
                        </div>
                        <template x-for="weekday in weekDays" :key="weekday">
                            <div class="w-48 p-2 mb-2 border rounded hover:cursor-pointer hover:bg-gray-200"
                                @click="dayClick(week.weekNo, weekday)">
                                <div x-text="weekday"></div>
                            </div>
                        </template>
                    </div>
              </template>
            </div>
            <div>Alpine Property "selectedWeekDay": <span class="ml-2 font-bold" x-text="selectedWeekDay"></span></div>
        </div>
    </div>
    <button class="bg-gray-700 text-white px-4 py-2 my-4" wire:click="toggleShowDay">Toggle Show Livewire
        Result
    </button>
    @if ($open)
        <div>
            <h2 class="text-lg font-bold mb-4">Result from Livewire:</h2>
            <div>Weekday from Livewire: {{ $selectedDay }}</div>
        </div>
    @endif
</div>
```
This simple code works with Livewire 2 but not with version 3.

**Comment by Enrica-r** [comment]
Next try. Instead of one Alpine data object in the main div, I tried have separate data object for the inner loop. Like this:

``` html
    <div x-data="{
        properties: [1, 2 ],
        function (): {}
    }">
        <template x-for="property in properties" :key="property">
                   <div x-text=week.weekNo></div>
                   <div x-data = "{
                            subproperties: [11, 12, 13],
                        }">
                            <template x-for="subpropery in suproerties" :key="suproperty">
                                    <div x-text="subpropery"></div>
                            </template>
                    </div>
         </template>
</div>
```
After calling a wire:click function, the `subproperties` is not defined. The error message is `livewire.js?id=11c49d7e:1256 Uncaught ReferenceError: weekday is not defined`.

So there is no chance to build Livewire independent a Blade component with Alpine using it within a Livewire fullpage component.

**Reply by sueboy19** [reply]
Template x-for inside only one root element.

```html
<template x-for="property in properties" :key="property">
                   <div x-text=week.weekNo></div>
                   <div x-data = "{
                            subproperties: [11, 12, 13],
                        }">
                            <template x-for="subpropery in suproerties" :key="suproperty">
                                    <div x-text="subpropery"></div>
                            </template>
                    </div>
         </template>
````

But this correct.

```html
<div class="grid grid-cols-[1fr_2fr]">
            <div class="">
                <template x-for="week in weeks" :key="week.weekNo">
                    <div>
                        <div class="flex">
                            <div>Week no.: </div>
                            <div x-text=week.weekNo></div>
                        </div>
                        <template x-for="weekday in weekDays" :key="weekday">
                            <div class="w-48 p-2 mb-2 border rounded hover:cursor-pointer hover:bg-gray-200"
                                @click="dayClick(week.weekNo, weekday)">
                                <div x-text="weekday"></div>
                            </div>
                        </template>
                    </div>
              </template>
            </div>
            <div>Alpine Property "selectedWeekDay": <span class="ml-2 font-bold" x-text="selectedWeekDay"></span></div>
        </div>
```

**Comment by gofish543** [comment]
In general I've noticed that Livewire and Alpine properties are becoming one request out of sync after the first Livewire request is made. This could be related

**Reply by roman-rr** [reply]
That's true. Sometimes some delay needed to make it sync. Still actual.

**Comment by Enrica-r** [comment]
Does nobody use Alpine with nested x-for loops and use this component in Livewire pages?

If yes, how did you solve the above issue? 

Without a fix I have to go back to Livewire 2 or change to another JS Framework for my component.

**Comment by fixcoderu** [comment]
I am faced with a similar problem of nested loops working.
Has anyone been able to solve this problem?

**Comment by aftabkhaliq86** [comment]
![image](https://github.com/livewire/livewire/assets/129963183/4b4cb42d-a6bb-4f14-b5de-16c957b945b2)

i habe the same issue when wire navigate and comes back to same page alpine expression  error

**Comment by N1ebieski** [comment]
I've tried @Enrica-r 's component on [wirebox](https://wirebox.app/b/41dz5) and it seems ```wire:ignore``` resolve this problem...

**Comment by Enrica-r** [comment]
@N1ebieski : I have found the solution with `wire:ignore` too. But if you try to push the toggle button in my example which uses a property from livewire, it will not work. The change of the property is ignored within the alpine component. A workaround could be to read the livewire property in alipine JS with `$wire` property. I didn't try it. 

In difference to V2 livewire V3 doesn't reintialize the JS. There is a similar problem with `wire:navigate`. I saw documentions and posts which solved this in a JS event `livewire:navigated` to reload or reinit the JS. To reinit the Alpine object in this event.

**Reply by N1ebieski** [reply]
_A workaround could be to read the livewire property in alipine JS with $wire property._

It's better to use ```entangle``` in such a case.

**Comment by roman-rr** [comment]
Facing the same problem a the many!!! part of the app. 
For me, at this moment, solution is to wait server-works done with `wire:loading` and prevent clicks if server work "in progress".
